### PR TITLE
perf: minimize cursor::position queries

### LINKED ISSUE #968: reduce the `\e[6n` spam to improve high-ping ssh scenarios
currently reedline sends a `\e[6n` to the terminal after every redraw (pressing any key).
this causes reedline to block until it gets the cursor-position back from the terminal.  

lets say you have a 500ms ping (Geo-stationary satellite internet, cross-continent, etc):
1. you send a single keypress lets say `a`
1. 500ms: pass and it arrives at the server
1. 1sec: reedline requests the cursor position (`\e[6n`)
    1. 500ms your terminal receives the request
    1. 500ms reedline recieves the response
1. 500ms: reedline sends the updates to your terminal

so it takes 2 seconds to render 1 character change.. and half of it is caused by the cursor-position-request.

i found this by running nu via `^nu | ^tee test.bin` and checking `test.bin` after every key-press in a separate terminal.

## References

discussions about reedline being slow over ssh:
* https://github.com/nushell/reedline/issues/914
* <https://discord.com/channels/601130461678272522/614593951969574961/1431608904009388042>
* <https://discord.com/channels/601130461678272522/614593951969574961/1409160306256511030>

### PR DESCRIPTION
**Pull request by reubeno**
# Skip redundant cursor::position queries in the repaint hot path

Reedline issues a CSI 6n / DSR (`cursor::position()`) on every paint to check whether `prompt_start_row` has drifted from the terminal's real cursor row. That's a synchronous round-trip; over high-latency transports like SSH it dominates prompt-redraw latency.

Within a `read_line` cycle reedline is the sole writer to the terminal and updates `prompt_start_row` itself as its rendering scrolls — so the drift check is unnecessary in the steady state.

A new `prompt_start_row_stale: bool` on `Painter` tracks freshness:

- Cleared after the cursor query in `initialize_prompt_position`.
- Set by paths that yield the tty or write content the painter doesn't  model (`handle_resize`, `print_external_message`, `paint_line`,  `open_editor`).
- Re-cleared on the next paint by a confirmed-no-drift query.

The hot-path drift check skips the DSR when the cache is fresh. The change is wholly inside reedline, so the improvement carries to any consumer (e.g., Nushell, brush, custom REPLs) without changes on their side.

## Notes

I know that this is an area that may be challenging to test/validate. I'm open to any suggestions/feedback on how to ensure that this change does not regress existing functionality, while still improving the performance.

**_Full disclosure:_** as you can probably tell, I used Claude to identify and develop changes -- but I was heavily steering, and ultimately rewrote the code several times to address concerns that I had myself over maintainability challenges. I'm also placing this in draft first, with the hope to get early feedback on the approach and the maintainers' openness to considering this change. Any and all feedback is very much welcome, please :smile:

## Tradeoffs

- Each invalidating event trades the unconditional per-paint DSR for one DSR on the next paint; that's strictly better than the current per-paint cost, but not free on workloads heavy in external-printer batches or resizes.
- Terminals where `cursor::position()` fails see the same behavior as before: the drift check still runs and fails, with no optimization gain (and no regression).

## Measurements

Nushell 0.113.0 (HEAD) built against this branch's reedline vs against reedline `main`. PTY harness (portable-pty), 80×24, scripted 80 ms input pacing, 5 Enter presses per run, `--no-config-file --no-history`. Three samples per condition; the harness is deterministic and produced identical DSR counts across samples within each condition.

DSR counts:

| reedline | total DSRs (5 Enters) | per Enter |
| --- | --- | --- |
| `main` | 19 | 3.80 |
| this branch | 6 | 1.20 |

Same harness with `--simulated-rtt-ms 125` (each CSI 6n reply delayed 125 ms — comparable to a typical cross-region SSH round trip):

| reedline | total wall-clock (5 Enters) | per Enter |
| --- | --- | --- |
| `main` | ~2150 ms | ~430 ms |
| this branch | ~640 ms | ~127 ms |

The remaining ~1 DSR/cycle on this branch is the unavoidable one in `initialize_prompt_position`; reedline genuinely doesn't know where the host left the cursor before `read_line` is called.

[brush](https://github.com/reubeno/brush), a smaller `reedline`-based shell, shows the same shape (3.0 → 1.0 DSRs/cycle) — confirming the win is at the reedline layer, not shell-specific.

closes #968 

related #1010 
related #870

### CODE REVIEW

File: src/painting/painter.rs [hunk 1]
```diff
@@ -165,7 +165,15 @@ pub(crate) struct PromptLayout {
...
     stdout: W,
+    /// Row where the prompt starts on screen. `_stale = false` means
+    /// the cached value is consistent with the terminal as of the
+    /// last successful query or paint; `true` means something may
+    /// have moved the cursor or written content we don't model, so
+    /// the next `repaint_buffer` must re-query. Call
+    /// `invalidate_prompt_start_row` from any new code path that
+    /// violates this invariant.
     prompt_start_row: u16,
+    prompt_start_row_stale: bool,
```
Comment (Bahex) [inline src/painting/painter.rs]:
```suggestion
    prompt_start_row: Option<u16>,
```
Using `Option` instead of two separate values that are semantically linked would be better. Having `None` instead of checking a `_stale` field ensures it's not possible to accidentally use a stale value.

edit: Looks like there is at least one place where the stale value can be useful. In that case a custom `enum` would be better. Maybe something like:
```rust
enum CachedValue<T> {
    Clear,
    Stale(T),
    Fresh(T),
}
```

Comment (reubeno) [reply src/painting/painter.rs]: Thanks for the feedback, @Bahex. I'd considered an enum but wanted to keep the change simpler; since you're also thinking it would be an improvement, I've pushed changes to combine the fields.

I went with a custom enum so I could have some convenience helpers implemented on it, but I'm also open to a generic `CachedValue<T>` like you suggest above if that's what you'd recommend/prefer.

If you have time to look at the updated version, please let me know what you think.

**Comment by bhouse-nexthop** [comment]
Drive-by from a downstream consumer that runs reedline over long-haul SSH — we hit exactly this and independently traced the inter-prompt lag to the per-paint `cursor::position()`, so a strong +1 on the approach. This also resolves #968 (and helps #1010/#870); might be worth a `Closes #968` to link it.

Two things from reading the diff:

1. **`clear_screen` / `clear_scrollback` look unhandled.** Both `MoveTo(0, 0)` but don't update `prompt_start_row`, and the engine's `ClearScreen` / `ClearScrollback` arms (`engine.rs` ~872 / ~1107) don't invalidate. Today the next paint's DSR re-anchors; once that DSR is skipped on a `Verified` cache, wouldn't `Ctrl-L` re-anchor at the stale pre-clear row? Since the cursor is *known* to be at row 0 after a clear, `mark_verified(0)` there would be both correct and free (no DSR). A `paint → Ctrl-L → paint` test would pin it.
2. **The residual `initialize_prompt_position` DSR** you call out: for consumers that fully own their output and always hand control back at a known position (e.g. column 0 after a trailing newline), it'd be great to let the host *assert* the start position and skip even that one — an opt-in like `with_known_start_row(u16)` / a "host guarantees col 0" hint. Would you be open to a follow-up on top of this PR?

Thanks for doing this — the measurements are great.

**Comment by kpcyrd** [comment]
Thanks for working on this!

Is there any way we can avoid cursor::position in entirety? No other shell seems to be doing this (including the rustyline crate), they instead calculate the cursor position (because they themselves are the ones moving the cursor when printing the prompt).

I'm concerned that even one call might be too much ("if Rust is so fast and efficient, why does this feel slower than python").

**Comment by fdncred** [comment]
@kpcyrd Let's take one step at a time. I don't think reedline is able to manage all cursor positions at this time. I'd like to see reedline get there but let's take baby steps first.

@bhouse-nexthop Thanks for the input. I'd like to see your point 1 answered in this PR and 2 would work for a follow-up PR.

@reubeno Thanks for this cool PR. What are your thoughts on the comments? I'd like to move on with it soon if we can, so we have time to test it out.

**Comment by reubeno** [comment]
Thanks all for the replies!

I'm in agreement with @fdncred and @kpcyrd that it would be nice to get to a place where we can remove all or most `cursor::position()` calls *and* I'm on the same page with @fdncred that it's best to take things one step at a time.

Regarding @bhouse-nexthop's comments:

>clear_screen / clear_scrollback look unhandled. Both MoveTo(0, 0) but don't update prompt_start_row, and the engine's ClearScreen / ClearScrollback arms (engine.rs ~872 / ~1107) don't invalidate

I believe the existing PR logic is correct. Here's what I see...

The `ReedlineEvent::ClearScreen` and `ReedlineEvent::ClearScrollback` arms in `engine.rs` call into `Painter::clear_screen()` and `Painter::clear_scrollback()`, respectively. Each of those functions do issue a `.MoveTo(0, 0)`, and they both subsequently call `initialize_prompt_position()`, which in turn unconditionally calls `cursor::position()` and then updates `prompt_start_row`.

Because of this, I don't see a correctness issue there, but @bhouse-nexthop makes a good point that this could be optimized. With the call to `MoveTo(0, 0)` we know exactly where the cursor should be and could skip the cursor position check. With that said, this sounds like a change that could be made as a follow-on incremental improvement (i.e., separate PR) to keep things moving forward. That would be my preference, but I'm happy to incorporate it too if the reviewers/maintainers would prefer otherwise.

>for consumers that fully own their output and always hand control back at a known position (e.g. column 0 after a trailing newline), it'd be great to let the host assert the start position...

Agreed with everyone here that this is an interesting possibility and best suited as a separate PR after this one.

**Comment by fdncred** [comment]
We have about 3 weeks to test this and ensure it works as expected. So, let's try it. Thanks!

**Comment by reubeno** [comment]
> We have about 3 weeks to test this and ensure it works as expected. So, let's try it. Thanks!

Thanks, @fdncred! Please do holler at me if issues get reported related to this.

For what it's worth, I also spent time last night having Claude build out ~100 pty-based end-to-end UX tests against a test reedline app, taking inspiration from `neovim`'s tests. Thus far, the functional behavior they've exercised has been consistent before and after this change--but I'd like to add some more edge cases. (And of course, that's still no replacement for getting human eyes and bake time on this.)

If there's interest in tests like those, that's something I'd also be happy to share.

**Comment by fdncred** [comment]
ya, i've been looking for some way to test with a ptty. that would be very cool!

**Comment by reubeno** [comment]
> ya, i've been looking for some way to test with a ptty. that would be very cool!

@fdncred In case you're interested, [here's what claude put together for testing](https://github.com/reubeno/reedline/tree/pty-e2e-tests/tests/e2e). I'm not comfortable submitting a draft because I haven't actually reviewed the changes myself (yet).

One of the things I was most interested in is seeing if some of the active `reedline` issues could be reproduced with test cases; here was the result: https://github.com/reubeno/reedline/blob/pty-e2e-tests/tests/e2e/issue_repros.rs. An example:

image

Happy to chat about this some if interesting...

**Comment by fdncred** [comment]
ya, definitely very cool. I can't play with it in detail this second, but I want to look at it closer. 

I see it doesn't work on windows yet, that's ok for a first pass but we'd need to support it for windows specific issues like exists right now with wezterm. that issue is related to OSC133. so i'd like to be able to capture/test for particular ansi escape sequences that are supposed to be emitted by reedline. other than that, it's a definite YES! I'd love to see an official version of this landed at some point in the near future.
